### PR TITLE
[API-352] Code coverage creation and uploading to Codecov workflow

### PR DESCRIPTION
**Pull request by alisengul53**
Run from fork.
https://github.com/alisengul53/hazelcast-cpp-client/actions/runs/1142693780

Created report on Codecov(It doesn't show the correct results because in fork repo some AWS env veriables are not set)
https://codecov.io/github/alisengul53/hazelcast-cpp-client/commit/3ce581a61bbc3056197cb37a520f16cb7982f813

### CODE REVIEW

File: .github/workflows/CodeCoveraReport.yaml [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+        run: |
+          sudo ./scripts/install-boost.sh 1.76.0
+          
+      - name: Install Thrift
+        run: |
+          sudo ./scripts/install-thrift.sh 0.13.0
+         
+      - name: Setup Gcovr   
+        run: |
+            sudo apt install gcovr
```
Comment (yemreinci) [inline .github/workflows/CodeCoveraReport.yaml:38]: Can we add `gcovr` to the "Install Necessary Packages" step above and remove this step?

File: .github/workflows/CodeCoveraReport.yaml [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+          
+      - name: Build
+        env:
+          BUILD_DIR: build
+          COVERAGE: ON
+          HAZELCAST_ENTERPRISE_KEY: ${{ secrets.HAZELCAST_ENTERPRISE_KEY }}
```
Comment (yemreinci) [inline .github/workflows/CodeCoveraReport.yaml:44]: Building doesn't require enterprise key.
```suggestion
```

File: .github/workflows/CodeCoveraReport.yaml [hunk 3]
```diff
@@ -0,0 +1,77 @@
+name: Code coverage report to Codecov
+
+on:
+    push:
+      branches: 
+        - master
```
Comment (yemreinci) [inline .github/workflows/CodeCoveraReport.yaml:6]: minor indentation inconsistency
```suggestion
on:
  push:
    branches: 
      - master
```

File: .github/workflows/CodeCoveraReport.yaml [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+jobs:
+        
+  test_client:
```
Comment (yemreinci) [inline .github/workflows/CodeCoveraReport.yaml:9]:
```suggestion
  coverage:
```

File: .github/workflows/CodeCoveraReport.yaml [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+          sudo sh -c "echo 'core' > /proc/sys/kernel/core_pattern"
+          sudo sh -c "echo '1' > /proc/sys/kernel/core_uses_pid"
+          ./scripts/test-unix.sh       
+       
+      - name: Collect coverage files
+        run: |   
+              gcovr --xml-pretty -o cpp_coverage.xml                \
+                --delete                                            \
+                -r "$(pwd)/hazelcast" -e "$(pwd)/hazelcast/test"    \
+                "$(pwd)/build"
```
Comment (yemreinci) [inline .github/workflows/CodeCoveraReport.yaml:71]: indentation is inconsistent with other commands
```suggestion
        run: |   
          gcovr --xml-pretty -o cpp_coverage.xml                \
            --delete                                            \
            -r "$(pwd)/hazelcast" -e "$(pwd)/hazelcast/test"    \
            "$(pwd)/build"
```

**Review by yemreinci — COMMENTED** [review]
Leaved a few comments. Also, can you change the name of the workflow to sth like `coverage.yml` or `coverage-report.yml` to be consistent with the naming of other workflows?

Please also share a link to a run in your fork in the PR description, so we can check the logs and the coverage report created.

File: .github/workflows/CodeCoveraReport.yaml (hunk 1 above)
Comment (alisengul53) [reply .github/workflows/CodeCoveraReport.yaml:38]: make sense, fixing

**Comment by alisengul53** [comment]
> Leaved a few comments. Also, can you change the name of the workflow to sth like `coverage.yml` or `coverage-report.yml` to be consistent with the naming of other workflows?
> 
> Please also share a link to a run in your fork in the PR description, so we can check the logs and the coverage report created.

Yaml file name is set to 'coverage-report.yaml'. Run on fork link and coverage link on Codecov is added

File: .github/workflows/coverage-report.yaml [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+          java-version: 8
+            
+      - name: Checkout code           
+        uses: actions/checkout@v2
+        
+      - name: Install Necessary Packages
+        run: |
+          sudo apt-get update
+          sudo apt-get install -y net-tools libssl-dev gdb
+          sudo apt install gcovr
```
Comment (ihsandemir) [inline .github/workflows/coverage-report.yaml:27]: why not at the above line using `apt-get`?

**Comment by ihsandemir** [comment]
Why is the coverage showing too low at the provided [report](https://codecov.io/github/alisengul53/hazelcast-cpp-client/commit/3ce581a61bbc3056197cb37a520f16cb7982f813)

See our jenkins [run](http://jenkins.hazelcast.com/view/C++/job/cpp-linux-64-SHARED-Debug/) showing a line coverage of 88%.

Which coverage is being shown here? We consider the line coverage as the primary one.

**Comment by alisengul53** [comment]
> Why is the coverage showing too low at the provided [report](https://codecov.io/github/alisengul53/hazelcast-cpp-client/commit/3ce581a61bbc3056197cb37a520f16cb7982f813)
> 
> See our jenkins [run](http://jenkins.hazelcast.com/view/C++/job/cpp-linux-64-SHARED-Debug/) showing a line coverage of 88%.
> 
> Which coverage is being shown here? We consider the line coverage as the primary one.

Some AWS keys are not set in the fork repo, most probably that is why some of tests are not run and then coverage seems lower

File: .github/workflows/coverage-report.yaml (hunk 6 above)
Comment (alisengul53) [reply .github/workflows/coverage-report.yaml:27]: No specific reason actually, when we realized gcovr should be installed and the suggested installation command was 'apt install gcovr' in container. It was put as it is. Making it `apt-get`

**Comment by yemreinci** [comment]
> > Why is the coverage showing too low at the provided [report](https://codecov.io/github/alisengul53/hazelcast-cpp-client/commit/3ce581a61bbc3056197cb37a520f16cb7982f813)
> > See our jenkins [run](http://jenkins.hazelcast.com/view/C++/job/cpp-linux-64-SHARED-Debug/) showing a line coverage of 88%.
> > Which coverage is being shown here? We consider the line coverage as the primary one.
> 
> Some AWS keys are not set in the fork repo, most probably that is why some of tests are not run and then coverage seems lower

No, not that. It's way much lower AWS-things considered. I think Codecov doesn't count "partial coverage"s (shown in yellow on report pages). I am currently looking into that.

**Comment by yemreinci** [comment]
OK, most yellow lines seems to be due to "throw branches". gcovr reports those lines as partially covered because the case where the line could throw wasn't covered by the tests. I think we don't need to consider those cases and they can be disabled by passing `--exclude-throw-branches` to `gcovr`. 

I did a test run with `--exclude-throw-branches` on my fork, see [logs](https://github.com/yemreinci/hazelcast-cpp-client/runs/3361542885?check_suite_focus=true) and [report](https://codecov.io/gh/yemreinci/hazelcast-cpp-client/tree/64eadc459e7b3138ede63a622dd5a980bb6dff46/). Coverage is 80% but I also didn't run the AWS tests as my keys were broken, so I am hoping it will be similar to Jenkins if we run them as well.

@alisengul53, I'm adding the change as a review suggestion.

File: .github/workflows/coverage-report.yaml [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+          java-version: 8
+            
+      - name: Checkout code           
+        uses: actions/checkout@v2
+        
+      - name: Install Necessary Packages
+        run: |
+          sudo apt-get update
+          sudo apt-get install -y net-tools libssl-dev gdb
+          sudo apt-get install gcovr
```
Comment (yemreinci) [inline .github/workflows/coverage-report.yaml:27]: I think what @ihsandemir meant was this:
```suggestion
          sudo apt-get install -y net-tools libssl-dev gdb gcovr
```

File: .github/workflows/coverage-report.yaml [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+        run: |
+          ulimit -c unlimited
+          sudo sh -c "echo 'core' > /proc/sys/kernel/core_pattern"
+          sudo sh -c "echo '1' > /proc/sys/kernel/core_uses_pid"
+          ./scripts/test-unix.sh       
+       
+      - name: Collect coverage files
+        run: |   
+          gcovr --xml-pretty -o cpp_coverage.xml                \
+            --delete                                            \
```
Comment (yemreinci) [inline .github/workflows/coverage-report.yaml:65]:
```suggestion
            --delete                                            \
            --exclude-throw-branches                            \
```

**Review by yemreinci — APPROVED** [review]
Thanks!

**Comment by yemreinci** [comment]
@alisengul53 I have accidentally pushed my changes to your branch but took it back.

**Comment by yemreinci** [comment]
I did one more run, this time all tests enabled: https://github.com/yemreinci/hazelcast-cpp-client/runs/3362418945?check_suite_focus=true. 
Coverage is still lower than Jenkins (81%).

**Comment by ihsandemir** [comment]
> I did one more run, this time all tests enabled: https://github.com/yemreinci/hazelcast-cpp-client/runs/3362418945?check_suite_focus=true.
> Coverage is still lower than Jenkins (81%).

I assume both Jenkins and ga uses the result of the gcovr and I expect both to report the same result since the xml report should be the same for both. We need to make sure that it is `line coverage` as we take as the major coverage number. May be we can compare the xml reports and try to understand what is different. Do you have both xml files?

**Comment by yemreinci** [comment]
I pushed the fix to this PR. The reported coverage rate is correct now. See the workflow logs from my fork:
https://github.com/yemreinci/hazelcast-cpp-client/actions/runs/1697547070

See published Codecov report:
https://codecov.io/gh/yemreinci/hazelcast-cpp-client/tree/c2727af734825162e8cd990a402b91e994735362

I've also added a badge in the README. Check my fork to see how it looks:
https://github.com/yemreinci/hazelcast-cpp-client#readme

@ihsandemir, could you please review?

**Review by yemreinci — APPROVED** [review]

**Comment by ihsandemir** [comment]
> I pushed the fix to this PR. The reported coverage rate is correct now. See the workflow logs from my fork: https://github.com/yemreinci/hazelcast-cpp-client/actions/runs/1697547070
> 
> See published Codecov report: https://codecov.io/gh/yemreinci/hazelcast-cpp-client/tree/c2727af734825162e8cd990a402b91e994735362
> 
> I've also added a badge in the README. Check my fork to see how it looks: https://github.com/yemreinci/hazelcast-cpp-client#readme
> 
> @ihsandemir, could you please review?

@yemreinci 
I am looking at [this](https://app.codecov.io/gh/yemreinci/hazelcast-cpp-client/blob/c2727af734825162e8cd990a402b91e994735362/hazelcast/include/hazelcast/client/serialization/serialization.h) file in your result, and I see some lines are not marked in any color, what do they mean? E.g. line 128

**Comment by yemreinci** [comment]
> > I pushed the fix to this PR. The reported coverage rate is correct now. See the workflow logs from my fork: https://github.com/yemreinci/hazelcast-cpp-client/actions/runs/1697547070
> > See published Codecov report: https://codecov.io/gh/yemreinci/hazelcast-cpp-client/tree/c2727af734825162e8cd990a402b91e994735362
> > I've also added a badge in the README. Check my fork to see how it looks: https://github.com/yemreinci/hazelcast-cpp-client#readme
> > @ihsandemir, could you please review?
> 
> @yemreinci I am looking at [this](https://app.codecov.io/gh/yemreinci/hazelcast-cpp-client/blob/c2727af734825162e8cd990a402b91e994735362/hazelcast/include/hazelcast/client/serialization/serialization.h) file in your result, and I see some lines are not marked in any color, what do they mean? E.g. line 128
> 
> Looks good other than this.

Line 128 of serialization.h is a function definition. It's not actually executable code so it's not counted as a code line and left uncolored. If you look at the Jenkins coverage reports that we already have, you'll see that they have uncolored lines as well:
http://jenkins.hazelcast.com/view/C++/job/cpp-pr-builder/1192/cobertura/include_hazelcast_client_serialization/serialization_h/

**Review by ihsandemir — APPROVED** [review]

**Pull request merged**